### ISSUE TITLE
Excessive ANR Errors on HUAWEI devices

### ISSUE THREAD

**Issue by stephenruda** [bugsnag/bugsnag-android#467]
Hi, we just shipped a new release of our app with the latest version (v4.13.0) of the Bugsnag library.

We are noticing a lot of ANR exceptions coming exclusively from HUAWEI devices.  The stacktraces have been relatively unhelpful.  I noticed some of them happening when an activity calls onResume() and I noticed some others happening at random after the app has been in the background for a while (1 hour+).

My thought is that HUAWEI protected apps / phone manager features are doing something menacing that slows down or stops processing which leads to these ANR errors popping up.  I don't see any ANR errors popping up on the Play Store Console and no other devices have similar ANR errors.

I don't have a HUAWEI device to test on.  I am just going off what I see on our bugsnag dashboard from our production users.

For now we are just going to ignore these errors.  Maybe someone else can shed some light on this issue or report similar results.

Thanks!

**Comment by MikeKasperlik** [comment]
We see this one too. Our play console report shows no signs of ANRs. 

We have a service running in background in a separate process which does not require the app to be open. Maybe this is related?

**Comment by stephenruda** [comment]
Yeah, maybe it is related to services.

Our app has a bunch of services that are constantly being run.  Some foreground services and some background services.  Unfortunately, the lines of code that the ANRs are being generated on are completely random so there isn't really anything for me to go on.

I ended up just disabling ANR checking and I will just use the Play Store to do it.

**Comment by Lomaas** [comment]
Have this one too. We have a background service running and we are using bugsnag. No ANR reports in Play

**Comment by wkoutre** [comment]
Chiming in here to say almost the same, though using `react-native-bugsnag`:

- using `react-native-bugsnag@2.17.0`
- no ANR reports in Google Play Dev Console
- many ANR reports in Bugsnag for HUAWEI devices, specifically when background calls `onResume()` and also after the app has been in the background for a while, as @stephenruda mentioned

**Comment by emilioicai** [comment]
Same here. In my case I see the ANR reports in Google Play Dev Console

**Comment by LuisBurgos** [comment]
Having this one too. 

- No ANR reports in Google Play Dev Console 😦 
- Many ANR reports in BugSnag for HUAWEI devices, specifically when SCREEN_OFF and SCREEN_ON events are dispatched by the system. We do not have any receiver o listener for those events.

**Comment by snmaynard** [comment]
We're looking at improving the accuracy of our ANR detection, and also looking at Huawei devices specifically. Thanks for all the information provided in this thread, it helps us improve the product and get to the bottom of what you are all seeing!

**Comment by ywongweb** [comment]
Reverted my bugsnag back to `2.12.4`, no ANRs on that release so far.

**Comment by RogerParis** [comment]
We also got this issue and in my opinion it is Huawei and Samsung skins task managers killing the app silently when it is doing too much work in the background. The skins kill the app silently so no report is shown in google play but bugsnag actually cathes it. Which in my opinion is good. But it is really hard to reproduce.

**Comment by jppasa** [comment]
Also seeing excessive ANRs from Huawei devices. ANRs are not showing in Play console. 

Is there any way to, at least, deactivate ANR reporting for Huawei devices so my dashboard (and my email inbox) is no flooded from this reports?

**Comment by snmaynard** [comment]
You can discard ANRs from within the dashboard - but that will do it for all devices. You can also filter ANRs from the view by adding a 'errorClass:!ANR' filter in the UI.

We have our Huawei devices but are struggling to reproduce it. We have some theories that #482 may solve this but we aren't certain as we can't reproduce it.

We have also in parallel tried to come up with a more reliable detection method and we think we have a working implementation of that - and that should be coming in the next few weeks (hopefully sooner). cc @fractalwrench and @kattrali as the people looking into this!

**Comment by fractalwrench** [comment]
We have released v4.14.1 which contains the changes in #482. We believe that this may help reduce false positives in ANRs but aren't certain that it will fully resolve the issue, so would appreciate any feedback. As @snmaynard mentioned, longer-term we're investigating using a SIGQUIT handler to capture ANRs which should be more reliable.

**Comment by CodyBrouwers** [comment]
I'm using https://github.com/bugsnag/bugsnag-react-native in my ReactNative app which uses this package for the Android side and am seeing this same issue for Huawei phones. To disable ANR detection for Huawei phones for now I've instantiated the library differently depending on the Android build manufacturer. As I'm not an Android developer by any means, would love some feedback if this should be alright and I assume anyone else could do something similar directly with this library as described [here](https://docs.bugsnag.com/platforms/android/sdk/configuration-options/#setdetectanrs)

```java
@Override
public void onCreate() {
  super.onCreate();

  if ("huawei".equalsIgnoreCase(android.os.Build.MANUFACTURER)) {
    Configuration config = new Configuration("my-api-key");
    config.setDetectAnrs(false);
    BugsnagReactNative.startWithConfiguration(this, config);
  } else {
    // Bugsnag React Native pulls the api key from the AndroidManifest if not provided here I believe.
    BugsnagReactNative.start(this);
  }
}
```

**Comment by Lomaas** [comment]
@fractalwrench will 4.41.1 fixes arrive to bugsnag-react-native also?

**Comment by fractalwrench** [comment]
We've released a fix for this in v4.15.0, which uses a SIGQUIT handler to capture whenever the ANR dialog is shown. This should reduce false positives compared to the previous detection method.

To enable the newest ANR detection, please update to v4.15.0 of bugsnag-android and v4.4.0 of the bugsnag gradle plugin, and set `config.detectNdkCrashes` to true. We'd appreciate any feedback on how this compares to the previous implementation.

A release containing this fix for bugsnag-react-native will follow shortly.
